### PR TITLE
Chk 1194 retry on refund error

### PR DESCRIPTION
**Pull request by pietro-tota**
Depends on pagopa/pagopa-ecommerce-commons#53

#### List of Changes
 Add logic to perform retry on refund.
If the refund process cannot be completed successfully then a refund retry event is enqueue into the refund retry queue.
A new queue consumer has been added that receive events from the above queue and retry refund process for the input transaction.
If the refund has been processed successfully then the transaction is updated to `REFUNDED` or a new refund retry event is enqueue.
This process can go up until an attempt to perform refund for the transaction is available (configurable).
Each retry event is sent with an increasing visibility timeout so that each attempts is performed with a major delay compared to the previous one.
The delay is calculated as a base offset multiplied for the retry count so, giving a base offset of 10 seconds the following attempts will be performed:
| # attempt | delay |
|----------|--------|
| 1              | 10         |
| 2              | 20         |
| 3             | 30         |

#### Motivation and Context
Those modifications are needed in order to perform retry on refund

#### How Has This Been Tested?
Junit tests + local tests using ecommerce local

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by giovanniberti — COMMENTED** [review]
(I still have to finish reading the PR btw)
🚲: I would store the retry counter in reverse, i.e. starting from `maxAttempts` and checking whether the current attempt is equal to zero. This way if the retry mechanism gets stuck we can accurately pinpoint where are we in the retry process even in the face of configuration changes

**Comment by pietro-tota** [comment]
> (I still have to finish reading the PR btw) 🚲: I would store the retry counter in reverse, i.e. starting from `maxAttempts` and checking whether the current attempt is equal to zero. This way if the retry mechanism gets stuck we can accurately pinpoint where are we in the retry process even in the face of configuration changes

Yes, I totally agree, great!

File: src/main/kotlin/it/pagopa/ecommerce/scheduler/services/eventretry/RetryEventService.kt [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+              transaction.clientId,
+              transaction.creationDate.toString()))
+          .flatMap { Mono.just(event) }
+      }
+
+  private fun enqueueMessage(event: E, visibilityTimeoutMillis: Long): Mono<Void> =
+    Mono.just(event).flatMap { eventToSend ->
+      queueAsyncClient
+        .sendMessageWithResponse(
+          BinaryData.fromObject(eventToSend), Duration.ofMillis(visibilityTimeoutMillis), null)
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/scheduler/services/eventretry/RetryEventService.kt:89]:
```suggestion
          BinaryData.fromObject(eventToSend), Duration.ofMillis(visibilityTimeoutMillis), timeToLive = null)
```

File: src/main/kotlin/it/pagopa/ecommerce/scheduler/services/eventretry/RetryEventService.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        .flatMap {
+          logger.info(
+            "Event: [$event] successfully sent with visibility timeout: [${it.value.timeNextVisible}] ms to queue: [${queueAsyncClient.queueName}]")
+          Mono.empty<Void>()
+        }
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/scheduler/services/eventretry/RetryEventService.kt:94]:
```suggestion
        .doOnNext {
          logger.info(
            "Event: [$event] successfully sent with visibility timeout: [${it.value.timeNextVisible}] ms to queue: [${queueAsyncClient.queueName}]")
        }
```

File: src/main/kotlin/it/pagopa/ecommerce/scheduler/services/eventretry/RetryEventService.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+  fun enqueueRetryEvent(baseTransaction: BaseTransaction, retriedCount: Int): Mono<Void> {
+    val retryEvent =
+      buildRetryEvent(baseTransaction.transactionId, TransactionRetriedData(retriedCount + 1))
+    return Mono.just(retryEvent)
+      .filter { it.data.retryCount <= maxAttempts }
+      .switchIfEmpty(
+        Mono.error(
+          RuntimeException(
+            "No attempts left for retry event $retryEvent for transaction id: ${baseTransaction.transactionId}")))
```
[the remark this answers is not in the document]
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/scheduler/services/eventretry/RetryEventService.kt]: fixed with https://github.com/pagopa/pagopa-ecommerce-scheduler-service/pull/34/commits/929f57288b222581069da5e19a467da0a9149ab4

File: src/main/kotlin/it/pagopa/ecommerce/scheduler/services/eventretry/RetryEventService.kt (hunk 2 above)
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/scheduler/services/eventretry/RetryEventService.kt]: fixed with https://github.com/pagopa/pagopa-ecommerce-scheduler-service/pull/34/commits/929f57288b222581069da5e19a467da0a9149ab4

File: src/main/kotlin/it/pagopa/ecommerce/scheduler/services/eventretry/RetryEventService.kt [hunk 4]
```diff
@@ -32,13 +31,15 @@ abstract class RetryEventService<E>(
...
+      .doOnSuccess {
+        if (it == null) {
+          logger.info(
+            "No attempts left for retry event $retryEvent for transaction id: ${baseTransaction.transactionId}")
+        }
+      }
       .flatMap { storeEventAndUpdateView(baseTransaction, it, newTransactionStatus()) }
       .flatMap {
-        enqueueMessage(it, TimeUnit.SECONDS.toMillis((retryOffset * it.data.retryCount).toLong()))
+        enqueueMessage(it, Duration.ofSeconds((retryOffset * it.data.retryCount).toLong()))
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/scheduler/services/eventretry/RetryEventService.kt:42]: ❤️

File: helm/values-dev.yaml [hunk 5]
```diff
@@ -84,11 +85,14 @@ microservice-chart:
     MONGO_SSL_ENABLED: "true"
     ECOMMERCE_STORAGE_QUEUE_ACCOUNT_NAME: pagopadweuecommercesa
     ECOMMERCE_STORAGE_QUEUE_ENDPOINT: https://pagopadweuecommercesa.queue.core.windows.net/pagopadweuecommercesa
+    REFUND_RETRY_EVENT_OFFSET_SECONDS: 600
```
Comment (giovanniberti) [inline helm/values-dev.yaml:88]: 🚲: OFFSET -> INTERVAL / BASE_INTERVAL / INITIAL_INTERVAL

File: src/test/kotlin/it/pagopa/ecommerce/scheduler/queues/TransactionRefundRetryQueueConsumerTest.kt [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    StepVerifier.create(
+        transactionRefundRetryQueueConsumer.messageReceiver(
+          BinaryData.fromObject(refundRetriedEvent).toBytes(), checkpointer))
+      .verifyComplete()
+
+    /* Asserts */
+    verify(checkpointer, times(1)).success()
+    verify(paymentGatewayClient, times(1)).requestRefund(any())
+    val expectedRefundEventStatuses = listOf(TransactionEventCode.TRANSACTION_REFUNDED_EVENT)
+    val viewExpectedStatuses = listOf(TransactionStatusDto.REFUNDED)
```
Comment (giovanniberti) [inline src/test/kotlin/it/pagopa/ecommerce/scheduler/queues/TransactionRefundRetryQueueConsumerTest.kt:115]: Unused?

File: helm/values-dev.yaml (hunk 5 above)
Comment (pietro-tota) [reply helm/values-dev.yaml:88]: I like BASE_INTERVAL just because this is the base from which derive the effective visibility timeout of the event

Comment (pietro-tota) [reply helm/values-dev.yaml]: Fixed with https://github.com/pagopa/pagopa-ecommerce-scheduler-service/pull/34/commits/4e7ae58b849300572ed71d7b0df8569500cc53c6
Contextually open PR on local for system property refactoring
https://github.com/pagopa/pagopa-ecommerce-local/pull/48

**Review by giovanniberti — APPROVED** [review]
🚀

**Pull request merged**